### PR TITLE
[Merged by Bors] - Doc/module style doc blocks for examples

### PR DESCRIPTION
**Pull request by themasch**
# Objective

Provide a starting point for #3951, or a partial solution. 
Providing a few comment blocks to discuss, and hopefully find better one in the process. 

## Solution

Since I am pretty new to pretty much anything in this context, I figured I'd just start with a draft for some file level doc blocks. For some of them I found more relevant details (or at least things I considered interessting), for some others there is less. 

## Changelog

- Moved some existing comments from main() functions in the 2d examples to the file header level
- Wrote some more comment blocks for most other 2d examples

TODO: 
- [x] 2d/sprite_sheet, wasnt able to come up with something good yet 
- [x] all other example groups...

Also: Please let me know if the commit style is okay, or to verbose. I could certainly squash these things, or add more details if needed. 
I also hope its okay to raise this PR this early, with just a few files changed. Took me long enough and I dont wanted to let it go to waste because I lost motivation to do the whole thing. Additionally I am somewhat uncertain over the style and contents of the commets. So let me know what you thing please.

### CODE REVIEW

**Comment by IceSentry** [comment]
Your commit style is fine, we always squash the PR in a single commit using the PR title+description as the commit message so it doesn't matter how you commit while working on it.

File: examples/2d/contributors.rs
Comment (IceSentry) [inline examples/2d/contributors.rs:1]: Personally, I would prefer if comments explaining a specific system were directly on that system and not all at the top. I think the top level doc should just be a short explanation of what it is. I would start with having the same short description found in the examples README

Comment (alice-i-cecile) [reply examples/2d/contributors.rs:1]: Completely agree here :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Hey, thanks! I agree with IceSentry: we can keep the "module-level" comments pretty terse to start.

More detailed explanation (such as "why do we add `LogPlugin`) belongs with that code, because it's a) more helpful to users there and b) less likely to go out of sync.

**Comment by themasch** [comment]
First, thanks for you fast and helpful feedback! I really dont know what I was thinking when doing the `contributors` comment/novel.

> Your commit style is fine, we always squash the PR in a single commit using the PR title+description as the commit message so it doesn't matter how you commit while working on it.

Great, thanks!

> I think the top level doc should just be a short explanation of what it is. I would start with having the same short description found in the examples README

> I agree with IceSentry: we can keep the "module-level" comments pretty terse to start.

I agree that especially the `contributors` one is way to big, and really need to be split up. But on the other hand, I think that the description in the examples README list is a bit to terse for a file doc. I think, especially in examples, a comment should add a little bit more than just "sprite.rs renders a sprite".

Maybe a way between those extremes would be useful? Do you thing the comments for [`mesh2d.rs`](https://github.com/bevyengine/bevy/pull/4438/files#diff-e1fbb90f8e1e36cbbd74cd4723316df87e0493c1c109153e6605fbd664a789b4) and [`mesh2d_manual.rs`](https://github.com/bevyengine/bevy/pull/4438/files#diff-d0f56e34d21a5a9a91b9d54ab7f777f936f77d7e5da2ed6e8f9af87447078474) are okay? 

I will rework these by (re-)moving details and keeping the file level doc block terse.

**Comment by alice-i-cecile** [comment]
> Maybe a way between those extremes would be useful? Do you thing the comments for [mesh2d.rs](https://github.com/bevyengine/bevy/pull/4438/files#diff-e1fbb90f8e1e36cbbd74cd4723316df87e0493c1c109153e6605fbd664a789b4) and [mesh2d_manual.rs](https://github.com/bevyengine/bevy/pull/4438/files#diff-d0f56e34d21a5a9a91b9d54ab7f777f936f77d7e5da2ed6e8f9af87447078474) are okay?

I really like these comments :) They strike a nice balance, and provide some essential context on the examples.

**Comment by IceSentry** [comment]
Oh, yeah, I was just suggesting the README comments has a starting point. I definitely agree that more detail is necessary. My issue was specifically about the comments explaining specific systems not being directly next to the system,

**Comment by cart** [comment]
I'm on board with this general approach. Whenever y'all think this is ready, feel free to merge!

**Comment by themasch** [comment]
So, I think this is ready for review/feedback. All examples should have comments, I used the ones from the readme if I was not able to come up with something short and better (most of them), used existing comments where they existed.

**Comment by alice-i-cecile** [comment]
@bevyengine/docs-team can I get another set of eyes on this? :)

File: examples/stress_tests/many_cubes.rs
Comment (IceSentry) [inline examples/stress_tests/many_cubes.rs:3]: It seems like there's a missing word here. "To see a" or "To get a" or something like that.

File: examples/tools/scene_viewer.rs
Comment (IceSentry) [inline examples/tools/scene_viewer.rs:2]: I don't think the --  is necessary

**Review by IceSentry — APPROVED** [review]
I didn't look at everything in detail, but it looks good to me except a couple of nitpicks

File: examples/app/empty.rs
Comment (alice-i-cecile) [inline examples/app/empty.rs:1]: I don't love that this example exists, but that's not this PR's fault.

File: examples/app/headless.rs
Comment (alice-i-cecile) [inline examples/app/headless.rs:1]:
```suggestion
//! This example only enables a minimal set of plugins required for bevy to run.
//! In particular, it enables you to run the `App` without rendering, which can be very useful in tests or CI.
```

File: examples/app/headless_defaults.rs
Comment (alice-i-cecile) [inline examples/app/headless_defaults.rs:1]:
```suggestion
//! An application that runs with default plugins, but without an actual renderer
//! This can be very useful for integration tests or CI.
```

File: examples/reflection/trait_reflection.rs
Comment (alice-i-cecile) [inline examples/reflection/trait_reflection.rs:1]: This isn't terribly descriptive; but IMO we should tackle this as part of a push to properly document and explain bevy_reflect instead.

File: examples/scene/scene.rs
Comment (alice-i-cecile) [inline examples/scene/scene.rs:1]: This could really use a brief explanation of what a "scene" is in Bevy.

File: examples/shader/compute_shader_game_of_life.rs
Comment (alice-i-cecile) [inline examples/shader/compute_shader_game_of_life.rs:1]: What's a compute shader? ;)

File: examples/stress_tests/many_cubes.rs
Comment (alice-i-cecile) [reply examples/stress_tests/many_cubes.rs:3]:
```suggestion
//! To measure performance realistically, be sure to run this in release mode.
```

**Review by alice-i-cecile — COMMENTED** [review]
Overall LGTM. Great work; and we can build on these further.

In general, I think these could benefit from having a bit more "background explanation" about the topics they're covering, but I'm fine to address those in follow-up PRs.

IMO we should clean up the few typos, and then merge this.

File: examples/tools/scene_viewer.rs
Comment (themasch) [reply examples/tools/scene_viewer.rs:2]: You are correct. Just tested this, and - at least with 1.60 on windows - `cargo run` seems to pass additional paramters to the target binary. 
I'll remove all ` -- ` from `cargo run`s in the examples.

File: examples/app/empty.rs
Comment (themasch) [reply examples/app/empty.rs:1]: You want me to delete it? 
Not sure if thats something that should be done in this context?

Comment (alice-i-cecile) [reply examples/app/empty.rs:1]: I was trying to say that this should be done in a separate PR, yeah.

File: examples/shader/compute_shader_game_of_life.rs
Comment (themasch) [reply examples/shader/compute_shader_game_of_life.rs:1]: I added a sentence to add some context.

File: examples/scene/scene.rs
Comment (themasch) [reply examples/scene/scene.rs:1]: For me to add that, I'd need to understand that myself 😅. From what I understand, a `Scene` is a newtype wrapper around a `World` so we can put complete worlds - for example loaded from gltf assets - into our main world. That works by copying all components from the scene to the world its spawned in when the command is executed. Therefor,  it does not really 'mount' other worlds into our app.world, it just uses the container to spawn a coupled set of components/entities into it?

So 'a `Scene` is a set of entites that can be loaded from an asset and added to our world'?.

**Comment by themasch** [comment]
Thanks for the quick feedback, I tried to address all points.

**Review by alice-i-cecile — APPROVED** [review]

File: examples/scene/scene.rs
Comment (alice-i-cecile) [reply examples/scene/scene.rs:1]: Yep, we can clean this up later :)

File: examples/2d/mesh2d_manual.rs
Comment (Nilirad) [inline examples/2d/mesh2d_manual.rs:1]:
```suggestion
//! This example shows how to manually render 2d items using "mid level render apis" with a custom
```

Double space.

File: examples/2d/mesh2d_manual.rs
Comment (Nilirad) [inline examples/2d/mesh2d_manual.rs:4]:
```suggestion
//! It doesn't use the [`Material2d`] abstraction, but changes the vertex buffer to include vertex color.
//! Check out the "mesh2d" example for simpler / higher level 2d meshes.
```

Period at the end of sentence.

File: examples/2d/rect.rs
Comment (Nilirad) [inline examples/2d/rect.rs:15]:
```suggestion
    // Since this `Sprite` is not generated from an image, we have to specify its size in `custom_size`.
```

Monospace consistency.

File: examples/2d/rotation.rs
Comment (Nilirad) [inline examples/2d/rotation.rs:1]:
```suggestion
//! Demonstrates rotating entities in 2D using quaternions.
```

Just my opinion, feel free to skip if using “with” is sufficient.

File: examples/2d/texture_atlas.rs
Comment (Nilirad) [inline examples/2d/texture_atlas.rs:2]:
```suggestion
//! individual sprites.
```

Sentence period.

File: examples/3d/3d_scene.rs
Comment (Nilirad) [inline examples/3d/3d_scene.rs:1]:
```suggestion
//! Simple 3D scene with basic shapes and lighting.
```

Sentence period.

Comment (Nilirad) [inline examples/3d/3d_scene.rs:1]: Overall, I would think about changing the description into something less generic, that makes the reader expect more what they are about to see, since the focus of the example is on the scene itself; maybe something like:

> A simple 3D scene with light shining over a cube sitting on a plane.

File: examples/3d/lighting.rs
Comment (Nilirad) [inline examples/3d/lighting.rs:1]:
```suggestion
//! Illustrates various lighting options in a simple scene.
```

Sentence period.

Comment (Nilirad) [inline examples/3d/lighting.rs:1]: I think the description should talk about the fact that there are used both different light types (point, directional, ...) and different colors, instead of simply stating “various lighting options”.

File: examples/3d/load_gltf.rs
Comment (Nilirad) [inline examples/3d/load_gltf.rs:1]:
```suggestion
//! Loads and renders a glTF file as a scene.
```

This is the capitalization that Khronos [uses](https://www.khronos.org/gltf/). If one is referring to the file extension instead, they should write “`.gltf` file” instead. It is ok to say “glTF file” since it is another wording for “file with glTF format”.

I also added the period.

File: examples/2d/text2d.rs
Comment (Nilirad) [inline examples/2d/text2d.rs:1]:
```suggestion
//! Shows text rendering with moving, rotating and scaling text.
```

Saying it is an example doesn't convey any new or useful information.

**Review by Nilirad — CHANGES_REQUESTED** [review]
I'm doing this by pieces since the PR is quite big. I reached the `load_gltf` example. I mostly did consistency checks and, in a couple of cases, I suggested a longer, more specific suggestion.

**Comment by themasch** [comment]
Thanks for the annotations, @Nilirad! I went through the list and added periods where they where missing in the header comment. I did not fix punctuation for _all existing_ comments (yet), as I think that would increase the pull request size even more.
If you want me to tackle this as well, let me know.

File: examples/3d/lighting.rs
Comment (themasch) [reply examples/3d/lighting.rs]: Yeah, sounds good. I'll try to come up with something more helpful!

**Comment by alice-i-cecile** [comment]
>  I did not fix punctuation for all existing comments (yet), as I think that would increase the pull request size even more.

I agree with this; we should address this in a follow-up PR.

**Comment by Nilirad** [comment]
If we only change the lines already added by the PR, the total count won't change though.

File: examples/app/headless_defaults.rs
Comment (Nilirad) [inline examples/app/headless_defaults.rs:1]:
```suggestion
//! An application that runs with default plugins, but without an actual renderer.
```

File: examples/app/plugin.rs
Comment (Nilirad) [inline examples/app/plugin.rs:1]:
```suggestion
//! Demonstrates the creation and registration of a custom plugin.
```

File: examples/app/plugin_group.rs
Comment (Nilirad) [inline examples/app/plugin_group.rs:2]:
```suggestion
//! Demonstrates the creation and registration of a custom plugin group.
//! [`PluginGroup`]s are a way to group sets of plugins that should be registered together.
```

File: examples/app/return_after_run.rs
Comment (Nilirad) [inline examples/app/return_after_run.rs:1]:
```suggestion
//! Shows how to return to the calling function after a windowed Bevy app has exited.
```

**Review by Nilirad — CHANGES_REQUESTED** [review]
Day two. Still a long way to go. Some descriptions just look like longer version of the file name, but it's something that needs to be adjusted in a (or multiple) following PR.

File: examples/3d/update_gltf_scene.rs
Comment (themasch) [reply examples/3d/update_gltf_scene.rs:1]: ... I thought I looked for those... stupid me. Thanks for finding them.

File: examples/ecs/system_sets.rs
Comment (Nilirad) [inline examples/ecs/system_sets.rs:1]:
```suggestion
//! Shows how systems with a similar purpose can be grouped into sets.
//!
//! The systems in this example are organized in the following sets:
```

File: examples/ecs/system_sets.rs
Comment (Nilirad) [inline examples/ecs/system_sets.rs:22]:
```suggestion
//! Lastly a system with run criteria _done_ is used to exit the app.
```

Just few lines above, “criteria” has been used multiple times as singular, so I'm doing it here.

File: examples/ecs/timers.rs
Comment (Nilirad) [inline examples/ecs/timers.rs:1]:
```suggestion
//! Illustrates how `Timer`s can be used both as resources and components.
```

**Review by Nilirad — CHANGES_REQUESTED** [review]
I got to the ECS folder (included). I've been a little bit *audacious* with the ECS suggestions, since I'm not a complete newbie in that field. But I'm also not an expert, so feel free to discuss them if you have concerns.

File: examples/window/clear_color.rs
Comment (Nilirad) [inline examples/window/clear_color.rs:1]:
```suggestion
//! Shows how to set the solid color that is used to paint the window before the frame gets drawn.
//! 
//! Acts as background color, since pixels that are not drawn in a frame remain unchanged.
```

File: examples/window/low_power.rs
Comment (Nilirad) [inline examples/window/low_power.rs:1]:
```suggestion
//! This example illustrates how to run a winit window in a reactive, low power mode.
//! 
```

File: examples/window/multiple_windows.rs
Comment (Nilirad) [inline examples/window/multiple_windows.rs:1]:
```suggestion
//! Uses two windows to visualize a 3D model from different angles.
```

File: examples/window/scale_factor_override.rs
Comment (Nilirad) [inline examples/window/scale_factor_override.rs:1]:
```suggestion
//! This example illustrates how to override the window scale factor imposed by the
//! operating system.
```

File: examples/window/transparent_window.rs
Comment (Nilirad) [inline examples/window/transparent_window.rs:2]:
```suggestion
//! Shows how to display a window in transparent mode.
//! 
//! This feature works as expected depending on the platform. Please check the
//! [documentation](https://docs.rs/bevy/latest/bevy/prelude/struct.WindowDescriptor.html#structfield.transparent)
//! for more details.
```

File: examples/window/window_settings.rs
Comment (Nilirad) [inline examples/window/window_settings.rs:1]:
```suggestion
//! Illustrates how to change window settings and shows how to affect
//! the mouse pointer in various ways.
```

File: examples/shader/shader_material.rs
Comment (Nilirad) [inline examples/shader/shader_material.rs:1]:
```suggestion
//! Illustrates creating a custom material and a shader that uses it.
```

File: examples/README.md
Comment (Nilirad) [inline examples/README.md:246]:
```suggestion
`shader_material` | [`shader/shader_material.rs`](./shader/shader_material.rs) | Illustrates creating a custom material and a shader that uses it.
```

File: examples/README.md
Comment (Nilirad) [inline examples/README.md:248]:
```suggestion
`shader_material_screenspace_texture` | [`shader/shader_material_screenspace_texture.rs`](./shader/shader_material_screenspace_texture.rs) | A custom shader sampling a texture with view-independent UV coordinates.
```

File: examples/README.md
Comment (Nilirad) [inline examples/README.md:285]:
```suggestion
`global_vs_local_translation` | [`transforms/global_vs_local_translation.rs`](./transforms/global_vs_local_translation.rs) | Illustrates the difference between direction of a translation in respect to local object or global object Transform.
```

**Review by Nilirad — CHANGES_REQUESTED** [review]
That was a long journey 😅.

**Note for maintainers:** I found that some changes of this PR corrected some code formatting errors. Does CI check examples' format?

**Comment by themasch** [comment]
Thanks again @Nilirad, I copied all your suggestions. At least I hope I got them all correct, at this point I doubt my ability to do even that. 
Sorry for this terrible attempt, I guess I should have left this to someone more capable. 

Anyhow, I rebased on the current main branch, as there was a merge conflict (2d/rect.rs got deleted). A few new examples have been added in the meantime, some of them without commets, so I tried to add at least a basic one. Basically longer versions of the file name 😕.

**Comment by Nilirad** [comment]
You did a great job with this PR! The amount of files made it a gigantic task, and that's normal to make some mistakes. That's why reviews are fundamental.

---

I'll do now a quick final pass to ensure there aren't misplaced descriptions. I'll ignore typos, grammar, punctuation or form so we can merge it ASAP.

**Review by Nilirad — APPROVED** [review]
It looks like there are no grave mistakes. Thank you for accomplishing this task @themasch, and congratulations for your first merged contribution! 🎉

**Pull request closed without merging**
